File: python/twoSumIIinputArrayIsSorted#167/try1.py
# This is the optimal solution
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left = 0
        right = len(numbers) - 1
        while True:
            numSum = numbers[left] + numbers[right]
            if numSum > target:
                right -= 1
            elif numSum < target:
                left += 1
            else:
                return [left + 1, right + 1]

# Time: O(n)
# Space: O(1)

File: c++/longestSubstringWithoutRepeatingCharacter#3/try1.cpp
// This is close to the optimal solution
class Solution {
public:
    int lengthOfLongestSubstring(string s) 
    {
        if(s.size() == 0)
        {
            return 0;
        }
        unordered_map<char, int> charCount;
        int maxLength = 0;

        for(int left = 0, right = 0; right < s.size(); right++)
        {
            charCount[s[right]]++;
            while(charCount[s[right]] > 1)
            {
                charCount[s[left]]--;
                left++;
            }

            maxLength = max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

// Time: O(n)
// Space: O(n)
